src/bin/cargo/commands/remove.rs

```diff
@@ -154,28 +254,72 @@ fn gc_workspace(workspace: &Workspace<'_>, dependencies: &[String]) -> CargoResu
     Ok(())
 }
 
-/// Get whether or not a dependency is depended upon in a workspace.
-fn dep_in_workspace(dep: &str, members: &[LocalManifest]) -> bool {
-    members.iter().any(|manifest| {
-        manifest.get_sections().iter().any(|(_, table)| {
-            table
-                .as_table_like()
-                .unwrap()
-                .get(dep)
-                .and_then(|t| t.get("workspace"))
-                .and_then(|v| v.as_bool())
-                .unwrap_or(false)
-        })
-    })
+/// Check whether or not a package ID spec matches any dependencies.
+fn spec_has_match(
+    spec: &PackageIdSpec,
+    dependencies: &[Dependency],
+    config: &Config,
+) -> CargoResult<bool> {
+    for dep in dependencies {
+        match dep.source_id(config)? {
+            MaybeWorkspace::Other(source_id) => {
+                if spec.name().as_str() != &dep.name {
+                    continue;
+                }
```

Comment:
nit: I'd move this to the outer-most scope or maybe even make it a `filter` on `dependencies.iter()`